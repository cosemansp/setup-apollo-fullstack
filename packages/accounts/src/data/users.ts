import faker from 'faker';
import { User } from '../domain/user';

let users: User[] = [];

function generateUsers(count) {
  const users: User[] = [];
  for (let i = 0; i < count; i++) {
    // eslint-disable-line
    let firstName;
    let imageUrl;
    const random = faker.random.number(2);
    const lastName = faker.name.lastName();
    if (random === 1) {
      firstName = faker.name.firstName(1);
      imageUrl = `http://api.randomuser.me/portraits/lego/${faker.random.number(9)}.jpg`;
      // imageUrl = `https://api.adorable.io/avatars/400/${firstName}-${lastName}`;
    } else {
      firstName = faker.name.firstName(0);
      imageUrl = `http://api.randomuser.me/portraits/lego/${faker.random.number(9)}.jpg`;
      // imageUrl = `https://api.adorable.io/avatars/400/${firstName}-${lastName}`;
    }
    users.push({
      id: 1000 + i,
      firstName,
      lastName,
      age: faker.random.number(100),
      email: `${firstName}.${lastName}@${faker.internet.domainName()}`.toLowerCase(),
      image: imageUrl,
      phone: faker.phone.phoneNumber(),
      company: faker.company.companyName(),
      address: {
        street: faker.address.streetName(),
        city: faker.address.city(),
        zip: faker.address.zipCode(),
      },
    });
  }
  return users;
}

export const clearUsers = () => {
  users = [];
};

export const seedUsers = (numberOfUsers) => {
  users = generateUsers(numberOfUsers);
};

export const getAllUsers = (): User[] => {
  return users;
};

export const getUser = (id): User => {
  return users.find((user) => user.id === id);
};

export const deleteUser = (user: User): User[] => {
  users = users.filter((item) => user.id !== item.id);
  return users;
};

export const addUser = (user: User): User => {
  users.push(user);
  return user;
};
