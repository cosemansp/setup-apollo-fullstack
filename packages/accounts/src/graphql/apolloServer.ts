import util from 'util';
import { ApolloServer } from 'apollo-server-express';
import { buildFederatedSchema } from '@apollo/federation';
import { GraphQLSchema } from 'graphql';
import { verify, decode } from 'jsonwebtoken';
import { makeExecutableSchema } from 'apollo-server';
import { resolvers } from './resolvers';
import { typeDefs } from './schema';
import { Context, JWTPayload } from './context';
import { logManager } from '../logManager';
import { UserDataSource } from '@/dataSources/userDataSource';
import config from '@/config';

const log = logManager.getLogger('root.graphql');

// create GraphqlSchema
let graphqlSchema: GraphQLSchema;
if (config.FEDERATED) {
  graphqlSchema = buildFederatedSchema([{ typeDefs, resolvers }]);
} else {
  graphqlSchema = makeExecutableSchema({
    typeDefs,
    resolvers,
    allowUndefinedInResolve: true,
  });
}

// we need to omit the dataSources because they are added later by the graphqlServer
type BaseContext = Omit<Context, 'dataSources'>;

export const server = new ApolloServer({
  schema: graphqlSchema,
  context: ({ res, req }): BaseContext => {
    // Base context
    const context: BaseContext = {
      req,
      res,
    };

    // Add user on context if we have an authorization header
    try {
      const authorization = req.headers.authorization;
      const token = authorization ? authorization.split(' ')[1] : '';
      if (token) {
        // if we have a token, verify it
        if (config.FEDERATED) {
          // when federated, let the gateway verify the token
          context.user = decode(token) as JWTPayload;
        } else {
          context.user = verify(token, config.ACCESS_TOKEN_SECRET) as JWTPayload;
        }
      }
    } catch (err) {
      // we don't throw an error here
      // the auth directives will handle the authorization
      log.warn('Failed to verify accessToken, ERR: ', err.message);
    }
    return context;
  },
  dataSources() {
    return {
      user: new UserDataSource(),
    };
  },
  formatError: (err) => {
    log.error('GRAPQH_ERROR >>>', util.inspect(err, false, 4, true /* enable colors */));
    return err;
  },
  playground: true,
  introspection: true,
});

export const schema = graphqlSchema;
