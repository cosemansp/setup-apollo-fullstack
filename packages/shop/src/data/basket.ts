// we can store multiple baskets
let baskets = [];

export const getOrCreateBasket = (checkoutID) => {
  let basket = baskets[checkoutID];
  if (!baskets[checkoutID]) {
    baskets[checkoutID] = [];
    basket = baskets[checkoutID];
    basket.push({
      id: 1,
      productId: 1,
      quantity: 1,
    });
    basket.push({
      id: 2,
      productId: 2,
      quantity: 4,
    });
  }
  return basket;
};

export const removeProductFromBaskets = (productId) => {
  for (let prop in baskets) {
    baskets[prop] = baskets[prop].filter((basket) => basket.productId !== productId);
  }
};

export const clearBasket = (checkoutID, refill = false) => {
  const previousBasket = getOrCreateBasket(checkoutID);
  baskets[checkoutID] = []; // clear basket
  if (refill) {
    baskets[checkoutID].push({
      id: 1,
      productId: 1,
      quantity: 1,
    });
    baskets[checkoutID].push({
      id: 2,
      productId: 2,
      quantity: 4,
    });
  }
  return previousBasket;
};
