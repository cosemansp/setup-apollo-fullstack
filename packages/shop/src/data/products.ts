import sampleProducts from './sampleProducts';

let products = [];

export const clearProducts = () => {
  products = [];
};

export const seedProducts = (count = 100) => {
  // start clean
  clearProducts();

  // copy from seed products
  for (let i = 0; i < count; i++) {
    products.push({
      ...sampleProducts[i],
    });
  }
  return products;
};

export const getAllProducts = () => {
  return products;
};

export const getProduct = (id) => {
  return products.find((product) => product.id === id);
};

export const deleteProduct = (product) => {
  products = products.filter((item) => product.id !== item.id);
  return products;
};

export const addProduct = (product) => {
  if (!product.price) {
    product.price = product.basePrice;
  }
  if (!product.stocked) {
    product.stocked = false;
  }
  products.push(product);
  return product;
};

export const addProducts = (products) => {
  products.forEach((product) => {
    this.addProduct(product);
  });
};
