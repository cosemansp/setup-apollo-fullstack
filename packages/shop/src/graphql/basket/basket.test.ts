import { createTestClient } from 'apollo-server-testing';
import gql from 'graphql-tag';
import { apolloServer } from '../apolloServer';
import { seedProducts } from '../../data/products';

const { query, mutate } = createTestClient(apolloServer as any);

const BASKET_QUERY = gql`
  query {
    basket(checkoutID: "abc") {
      checkoutID
      items {
        id
        product {
          id
          title
        }
        quantity
      }
    }
  }
`;

describe('My test', () => {
  beforeAll(() => {
    seedProducts();
  });

  test('dot this', async () => {
    const result = await query({ query: BASKET_QUERY });
    expect(result.data).toMatchSnapshot();
  });
});
