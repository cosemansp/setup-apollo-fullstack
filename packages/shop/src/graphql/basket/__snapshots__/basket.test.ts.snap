// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`My test dot this 1`] = `
Object {
  "basket": Object {
    "checkoutID": "abc",
    "items": Array [
      Object {
        "id": "1",
        "product": Object {
          "id": 1,
          "title": "pellentesque",
        },
        "quantity": 1,
      },
      Object {
        "id": "2",
        "product": Object {
          "id": 2,
          "title": "ut",
        },
        "quantity": 4,
      },
    ],
  },
}
`;
